### PR TITLE
Add JSON log parsing

### PR DESCRIPTION
**Pull request by DebakelOrakel**
## Checklist

- [x] Keep pull requests small so they can be easily reviewed.
- [x] Update the documentation.
- [x] Categorize the PR by setting a good title and adding one of the labels:
      `bug`, `enhancement`, `documentation`, `change`, `breaking`, `dependency`
      as they show up in the changelog
- [ ] Link this PR to related issues.

### CODE REVIEW

**Comment by DebakelOrakel** [comment]
Depends on #14

**Comment by simu** [comment]
Please finish and merge #14 first and rebase this afterwards, it's unclear which changes belong to this PR and which belong to #14 currently.

File: docs/modules/ROOT/pages/references/parameters.adoc [hunk 1]
```diff
@@ -78,6 +77,29 @@ clusterLogging:
 <1> Namespace to configure.
 <2> Name of the forwarder defined in `clusterLogForwarding.forwarder`.
 
+=== `clusterLogForwarding.json`
+
+[horizontal]
+type:: dictionary
+default:: see below
+
+If `json.enabled` is set to `true` the logging stack attempts to parse log messages to the json format.
```
Comment (simu) [inline docs/modules/ROOT/pages/references/parameters.adoc:86]:
```suggestion
If `json.enabled` is set to `true` the logging stack attempts to parse log messages to the JSON format.
```

File: docs/modules/ROOT/pages/references/parameters.adoc [hunk 2]
```diff
@@ -78,6 +77,29 @@ clusterLogging:
...
+[source,yaml]
+----
+clusterLogForwarding:
+  json:
+    enabled: false <1>
+    typekey: 'kubernetes.labels.logFormat' <2>
+    typename: 'nologformat' <3>
+----
+
+<1> By default json parsing is disabled.
```
Comment (simu) [inline docs/modules/ROOT/pages/references/parameters.adoc:97]:
```suggestion
<1> By default JSON parsing is disabled.
```

**Review by simu — APPROVED** [review]
Implementation looks good.

Does this also affect logs which are ingested into the default cluster-logging Elasticsearch? That's not very clear from the documentation.

PS: why are there 15 commits in this PR?

**Comment by DebakelOrakel** [comment]
> Does this also affect logs which are ingested into the default cluster-logging Elasticsearch? That's not very clear from the documentation.
It does affect all logs. It works like this:
* it tries to parse the `message` field to json and saves it in a new field `structured`
* if the pod (depending on configuration) has a label `logformat=apache` it creates a new index `app-apache-xxxxx` where the structured fields are indexed and searchable
* if it cant parse the `message` field no `structured` field is created
 
> PS: why are there 15 commits in this PR?
I was wondering that either, probably bc i checked out from the other branch...

**Comment by simu** [comment]
> > PS: why are there 15 commits in this PR?
> I was wondering that either, probably bc i checked out from the other branch...

Looking at the commits in more detail, it's probably because you used GitHub's "Update Branch" which just merges master into the feature branch instead of rebasing the feature branch on the latest master.

File: docs/modules/ROOT/pages/references/parameters.adoc [hunk 3]
```diff
@@ -78,6 +77,29 @@ clusterLogging:
...
+    enabled: false <1>
+    typekey: 'kubernetes.labels.logFormat' <2>
+    typename: 'nologformat' <3>
+----
+
+<1> By default JSON parsing is disabled.
+<2> The value of that field, if present, is used to construct the index name.
+<3> If `typekey` isn't set or its key isn't present, the value of this field is used to construct the index name.
+
+See the https://docs.openshift.com/container-platform/4.7/logging/cluster-logging-enabling-json-logging.html#cluster-logging-configuration-of-json-log-data-for-default-elasticsearch_cluster-logging-enabling-json-logging[OpenShift docs] for a detailed explanation.
```
Comment (corvus-ch) [inline docs/modules/ROOT/pages/references/parameters.adoc:101]:
```suggestion
See the https://docs.openshift.com/container-platform/4.8/logging/cluster-logging-enabling-json-logging.html#cluster-logging-configuration-of-json-log-data-for-default-elasticsearch_cluster-logging-enabling-json-logging[OpenShift docs] for a detailed explanation.
```

**Review by corvus-ch — CHANGES_REQUESTED** [review]
LGTM except for the git history. Please rebase.

**Comment by DebakelOrakel** [comment]
WTF... git, are you mad?

File: docs/modules/ROOT/pages/references/parameters.adoc (hunk 3 above)
Comment (DebakelOrakel) [reply docs/modules/ROOT/pages/references/parameters.adoc:102]: Ah, that should be latest.

**Review by corvus-ch — APPROVED** [review]

**Pull request merged**